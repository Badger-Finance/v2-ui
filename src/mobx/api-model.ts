/**
 * All API Model Objects should be directly in line with API interface model
 * objects to ensure data compatiblity. Interfaces located at link below:
 * https://github.com/axejintao/badger-api/tree/staging/src/interface
 */
import BigNumber from "bignumber.js";

export interface Geyser {
  emissions: Emission[],
}

export interface Token {
  address: string;
  name: string;
  symbol: string;
  decimals: number;
}

export interface Emission {
  token: Token,
  unlockSchedule: UnlockSchedule,
}

export interface UnlockSchedule {
	initialLocked: BigNumber,
	endAtSec: BigNumber,
	durationSec: BigNumber,
	startTime: BigNumber,
}

export interface TokenBalance {
  address: string;
  name: string;
  symbol: string;
  decimals: number;
  balance: number,
  value: number,
}

export interface ValueSource {
  name: string,
  apy: number,
  performance: Performance,
}

export interface Sett {
  name: string,
  asset: string,
  ppfs: number,
  value: number,
  apy: number,
  tokens: TokenBalance[],
  sources: ValueSource[],
  geyser?: Geyser,
}

export interface SettSummary {
  name: string,
  asset: string,
  value: number,
  tokens: TokenBalance[],
}
