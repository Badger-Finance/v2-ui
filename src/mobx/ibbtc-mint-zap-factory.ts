import { IbBTCMintZap } from './model/vaults/ibbtc-mint-zap';
import { RootStore } from './RootStore';
import addresses from 'config/ibBTC/addresses.json';
import { TokenZap } from './model/vaults/token-zap';
import { GeneralVaultZap } from './model/vaults/general-vault-zap';
import { RenVaultZap } from './model/vaults/ren-vault-zap';
import { Token } from '@badger-dao/sdk';

export class IbBTCMintZapFactory {
  private static tokenZaps: Record<string, TokenZap> = {};
  private static generalVaultZaps: Record<string, GeneralVaultZap> = {};
  private static renVaultZaps: Record<string, RenVaultZap> = {};

  static getIbBTCZap(store: RootStore, token: Token): IbBTCMintZap {
    if (addresses.mainnet.contracts.TokenZap.supportedTokens.includes(token.address)) {
      if (!this.tokenZaps[token.address]) {
        this.tokenZaps[token.address] = new TokenZap(store, token);
      }
      return this.tokenZaps[token.address];
    }

    if (addresses.mainnet.contracts.GeneralVaultZap.supportedTokens.includes(token.address)) {
      if (!this.generalVaultZaps[token.address]) {
        this.generalVaultZaps[token.address] = new GeneralVaultZap(store, token);
      }
      return this.generalVaultZaps[token.address];
    }

    if (addresses.mainnet.contracts.RenVaultZap.supportedTokens.includes(token.address)) {
      if (!this.renVaultZaps[token.address]) {
        this.renVaultZaps[token.address] = new RenVaultZap(store, token);
      }
      return this.renVaultZaps[token.address];
    }

    throw new Error(`Unrecognized peak type for token => ${token.symbol}`);
  }
}
