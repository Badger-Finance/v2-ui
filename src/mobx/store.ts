import { RouterStore } from 'mobx-router';
import UiState from './reducers';
import WalletStore from './stores/walletStore';
import ContractsStore from './stores/contractsStore';
import AirdropStore from './stores/airdropStore';
import RebaseStore from './stores/rebaseStore';
import RewardsStore from './stores/rewardsStore';
import IbBTCStore from './stores/ibBTCStore';
import TransactionsStore from './stores/transactionsStore';
import SettStoreV2 from './stores/settStoreV2';
import { NETWORK_LIST } from '../config/constants';
import HoneyPotStore from './stores/honeyPotStore';

export class RootStore {
	public router: RouterStore<RootStore>;
	public wallet: WalletStore;
	public uiState: UiState;
	public contracts: ContractsStore;
	public airdrops: AirdropStore;
	public rebase: RebaseStore;
	public rewards: RewardsStore;
	public ibBTCStore: IbBTCStore;
	public setts: SettStoreV2;
	public transactions: TransactionsStore;
	public honeyPot: HoneyPotStore;

	constructor() {
		this.router = new RouterStore<RootStore>(this);
		this.wallet = new WalletStore(this);
		this.contracts = new ContractsStore(this);
		this.airdrops = new AirdropStore(this);
		this.rebase = new RebaseStore(this);
		this.rewards = new RewardsStore(this);
		this.uiState = new UiState(this);
		this.ibBTCStore = new IbBTCStore(this);
		this.honeyPot = new HoneyPotStore(this);
		// RenVM transactions store.
		this.transactions = new TransactionsStore(this);
		this.setts = new SettStoreV2(this);

		this.walletRefresh();
	}

	async walletRefresh(): Promise<void> {
		const chain = this.wallet.network.name;
		const refreshData = [
			this.setts.loadAssets(chain),
			this.setts.loadPrices(chain),
			this.wallet.getGasPrice(),
			this.contracts.updateProvider(),
		];
		if (chain === NETWORK_LIST.ETH) {
			refreshData.push(this.setts.loadGeysers(chain));
			refreshData.push(this.rebase.fetchRebaseStats());
		} else {
			refreshData.push(this.setts.loadSetts(chain));
		}
		await Promise.all(refreshData);

		if (this.wallet.connectedAddress) {
			this.contracts.updateProvider();
			await this.wallet.getGasPrice();
			await this.contracts.fetchContracts();
			if (chain === NETWORK_LIST.ETH) {
				this.uiState.reduceRebase();
				this.ibBTCStore.init();
				this.rewards.fetchSettRewards();
				this.uiState.reduceTreeRewards();
				this.airdrops.fetchAirdrops();
				this.uiState.reduceAirdrops();
			}
			this.uiState.reduceStats();
		}
	}
}

const store = new RootStore();

export default store;
