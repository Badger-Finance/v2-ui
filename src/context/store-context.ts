import { createContext } from "react";
import { RootStore } from "../mobx/store";

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;