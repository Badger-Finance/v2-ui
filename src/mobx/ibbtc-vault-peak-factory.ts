import { IbbtcVaultPeak } from './model/vaults/ibbtc-vault-peak';
import { RootStore } from './store';
import addresses from 'config/ibBTC/addresses.json';
import { ZapPeak } from './model/vaults/zap-peak';
import { YearnPeak } from './model/vaults/yearn-peak';
import { BadgerPeak } from './model/vaults/badger-peak';
import { IbbtcOptionToken } from './model/tokens/ibbtc-option-token';

export class IbbtcVaultPeakFactory {
	static createIbbtcVaultPeakForToken(store: RootStore, token: IbbtcOptionToken): IbbtcVaultPeak {
		if (addresses.mainnet.contracts.ZapPeak.supportedTokens.includes(token.symbol)) {
			return new ZapPeak(store, token);
		}

		if (addresses.mainnet.contracts.yearnWBTCPeak.supportedTokens.includes(token.symbol)) {
			return new YearnPeak(store, token);
		}

		if (addresses.mainnet.contracts.BadgerSettPeak.supportedTokens.includes(token.symbol)) {
			return new BadgerPeak(store, token);
		}

		throw new Error(`Unrecognized peak type for token => ${token.symbol}`);
	}
}
