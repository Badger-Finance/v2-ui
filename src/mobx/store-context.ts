import { RootStore } from './store';
import { createContext } from 'react';

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;
