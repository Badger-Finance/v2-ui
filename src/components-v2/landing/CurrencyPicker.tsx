import React, { useContext } from 'react';
import { Select, MenuItem, makeStyles } from '@material-ui/core';
import { observer } from 'mobx-react-lite';
import { StoreContext } from '../../mobx/store-context';

const useStyles = makeStyles((theme) => ({
	currencyPicker: {
		height: '1.8rem',
		overflow: 'hidden',
		marginLeft: theme.spacing(1),
	},
}));

const CurrencyPicker = observer(() => {
	const classes = useStyles();
	const store = useContext(StoreContext);
	const {
		uiState: { currency, setCurrency },
		wallet: { network },
	} = store;

	return (
		<Select
			variant="outlined"
			value={currency}
			onChange={(v: any) => setCurrency(v.target.value)}
			className={classes.currencyPicker}
			style={{ marginTop: 'auto', marginBottom: 'auto' }}
		>
			<MenuItem value={'usd'}>USD</MenuItem>
			<MenuItem value={'cad'}>CAD</MenuItem>
			<MenuItem value={'btc'}>BTC</MenuItem>
			<MenuItem value={network.currency.toLowerCase()}>{network.currency}</MenuItem>
		</Select>
	);
});

export default CurrencyPicker;
