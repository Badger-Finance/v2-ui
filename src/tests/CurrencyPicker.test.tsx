import '@testing-library/jest-dom';

import { act, customRender, fireEvent, screen } from './Utils';

import CurrencyPicker from '../components-v2/landing/CurrencyPicker';
import React from 'react';
import { StoreProvider } from '../mobx/store-context';
import store from '../mobx/store';

describe('CurrencyPicker', () => {
	const testStore = store;

	test('Renders correctly', () => {
		const { container } = customRender(
			<StoreProvider value={testStore}>
				<CurrencyPicker />
			</StoreProvider>,
		);
		expect(container).toMatchSnapshot();
	});
	test('Opens menu upon click', async () => {
		customRender(
			<StoreProvider value={testStore}>
				<CurrencyPicker />
			</StoreProvider>,
		);
		await act(async () => {
			await fireEvent.mouseDown(screen.getByRole('button'));
		});
		expect(await screen.getByRole('presentation')).toMatchSnapshot();
	});
});
