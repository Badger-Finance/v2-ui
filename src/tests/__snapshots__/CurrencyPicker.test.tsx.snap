// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyPicker Opens menu upon click 1`] = `
<div
  class="MuiPopover-root-57"
  id="menu-"
  role="presentation"
  style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
>
  <div
    aria-hidden="true"
    style="z-index: -1; position: fixed; right: 0px; bottom: 0px; top: 0px; left: 0px; background-color: transparent;"
  />
  <div
    data-test="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiPaper-root-63 MuiMenu-paper-55 MuiPopover-paper-58 MuiPaper-elevation8-74 MuiPaper-rounded-64"
    style="opacity: 1; transform: none; min-width: 0; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
    tabindex="-1"
  >
    <ul
      class="MuiList-root-91 MuiMenu-list-56 MuiList-padding-92"
      role="listbox"
      tabindex="-1"
    >
      <li
        aria-disabled="false"
        aria-selected="true"
        class="MuiButtonBase-root-110 MuiListItem-root-99 MuiMenuItem-root-95 MuiMenuItem-selected-97 MuiMenuItem-gutters-96 MuiListItem-gutters-106 MuiListItem-button-107 MuiListItem-selected-109 MuiButtonBase-focusVisible-112 MuiListItem-focusVisible-101"
        data-value="usd"
        role="option"
        tabindex="0"
      >
        USD
        <span
          class="MuiTouchRipple-root-113"
        />
      </li>
      <li
        aria-disabled="false"
        class="MuiButtonBase-root-110 MuiListItem-root-99 MuiMenuItem-root-95 MuiMenuItem-gutters-96 MuiListItem-gutters-106 MuiListItem-button-107"
        data-value="cad"
        role="option"
        tabindex="-1"
      >
        CAD
        <span
          class="MuiTouchRipple-root-113"
        />
      </li>
      <li
        aria-disabled="false"
        class="MuiButtonBase-root-110 MuiListItem-root-99 MuiMenuItem-root-95 MuiMenuItem-gutters-96 MuiListItem-gutters-106 MuiListItem-button-107"
        data-value="btc"
        role="option"
        tabindex="-1"
      >
        BTC
        <span
          class="MuiTouchRipple-root-113"
        />
      </li>
      <li
        aria-disabled="false"
        class="MuiButtonBase-root-110 MuiListItem-root-99 MuiMenuItem-root-95 MuiMenuItem-gutters-96 MuiListItem-gutters-106 MuiListItem-button-107"
        data-value="eth"
        role="option"
        tabindex="-1"
      >
        ETH
        <span
          class="MuiTouchRipple-root-113"
        />
      </li>
    </ul>
  </div>
  <div
    data-test="sentinelEnd"
    tabindex="0"
  />
</div>
`;

exports[`CurrencyPicker Renders correctly 1`] = `
<div>
  <div
    class="MuiInputBase-root-28 MuiOutlinedInput-root-13 makeStyles-currencyPicker-1"
    style="margin-top: auto; margin-bottom: auto;"
  >
    <div
      aria-haspopup="listbox"
      class="MuiSelect-root-2 MuiSelect-select-3 MuiSelect-selectMenu-6 MuiSelect-outlined-5 MuiInputBase-input-39 MuiOutlinedInput-input-23"
      role="button"
      tabindex="0"
    >
      USD
    </div>
    <input
      aria-hidden="true"
      class="MuiSelect-nativeInput-12"
      tabindex="-1"
      value="usd"
    />
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root-46 MuiSelect-icon-8 MuiSelect-iconOutlined-11"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
    <fieldset
      aria-hidden="true"
      class="PrivateNotchedOutline-root-59 MuiOutlinedInput-notchedOutline-22"
      style="padding-left: 8px;"
    >
      <legend
        class="PrivateNotchedOutline-legend-60"
        style="width: 0.01px;"
      >
        <span>
          â€‹
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
